pkg/autoscaler/autoscaler.go

```diff
@@ -291,6 +294,106 @@ var _ = Describe("Autoscaler should", framework.LabelAutoscaler, Serial, func()
+				machineSet, err := framework.CreateMachineSet(client, machineSetParams)
+				Expect(err).ToNot(HaveOccurred(), "Failed to create MachineSet with 0 replicas")
+				machineSets = append(machineSets, machineSet)
+				cleanupObjects[machineSet.GetName()] = machineSet
+
+				framework.WaitForMachineSet(ctx, client, machineSet.GetName())
+
+				By(fmt.Sprintf("Creating a MachineAutoscaler backed by MachineSet %s/%s - min:%v, max:%v",
+					machineSet.GetNamespace(), machineSet.GetName(), 0, expectedReplicas))
+				asr := machineAutoscalerResource(machineSet, 0, expectedReplicas)
+				Expect(client.Create(ctx, asr)).Should(Succeed(), fmt.Sprintf("Failed to create MachineAutoscaler with min 0/max %v replicas", expectedReplicas))
+				cleanupObjects[asr.GetName()] = asr
+				arch := machineSetParams.Labels[framework.ArchLabel]
+				if arch != framework.Amd64 || workloadArch == "" {
+					// the workloadArch == "" condition ensures that workloadArch and expectedScaledMachineSet are set
+					// at least once. This is needed to ensure that the test does not fail when there is only one amd64 MachineSet
+					workloadArch = arch
+					expectedScaledMachineSet = machineSet
+				}
+			}
+			uniqueJobName := fmt.Sprintf("%s-scale-from-zero", workloadJobName)
+			By(fmt.Sprintf("Creating scale-out workload %s: jobs: %v, memory: %s", uniqueJobName,
+				expectedReplicas+1, workloadMemRequest.String()))
+			// Executing one additional replica to ensure that the workload does not triggers other scale-out events with
+			// other MachineSets not having the requested architecture.
+			workload := framework.NewWorkLoad(expectedReplicas+1, workloadMemRequest, uniqueJobName, autoscalingTestLabel, "",
+				corev1.NodeSelectorRequirement{
+					Key:      targetedNodeLabel,
+					Operator: corev1.NodeSelectorOpExists,
+				},
+				corev1.NodeSelectorRequirement{
+					Key:      "kubernetes.io/arch",
+					Operator: corev1.NodeSelectorOpIn,
+					Values:   []string{workloadArch},
+				})
+
+			cleanupObjects[workload.GetName()] = workload
+			Expect(client.Create(ctx, workload)).Should(Succeed(), "Failed to create scale-out workload %s", workloadJobName)
+
+			var condition bool
+			Consistently(func() bool {
+				condition = true
+				for _, machineSet := range machineSets {
+					if machineSet.GetName() == expectedScaledMachineSet.GetName() {
+						// Ignore the MachineSet with the architecture requested by the workload.
+						continue
+					}
+					ms, err := framework.GetMachineSet(ctx, client, machineSet.GetName())
+					Expect(err).ToNot(HaveOccurred(), "Failed to get MachineSet %s", machineSet.GetName())
+
+					By(fmt.Sprintf("Current replicas for %s are %d, expected %d.", machineSet.GetName(),
+						*ms.Spec.Replicas, 0))
+
+					condition = condition && *ms.Spec.Replicas == 0
+				}
+
+				return condition
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue(), "Some MachineSets scaled out unexpectedly.")
```

Comment:
Do we really need a WaitMedium for this? In our test utils we have a `RunCheckUntil` function that allows you to run a consistently style check, while waiting for some other condition, I think that may be what you want here

Reply:
Hi @JoelSpeed, which other condition could we use in this case?


Currently, the `Consistently` node continuously checks the machinesets that should not be scaled: they should never get a non-zero desired replicas value, regardless of the other being already set to 1.

The autoscaler should iterate and try to find a possible node group to scale even after the correct machineset is scaled to 1 (maximum number of replicas) because yet another pod is held in pending status for lack of available resources (both real and simulated).

Reply:
As per private discussion, we agreed on not having this "stronger" periodic check as it can still be weak. Resolving as fixed. Please help review the new code.